### ISSUE TITLE
Computing the best possible F1 score from a tree with ground truth labels

### ISSUE THREAD

**Issue by avasbr** [iesl/xcluster#3]
Hi,

First off, wanted to say thank you for providing the code for this technique - it's incredibly cool as a concept, and it was fun reading through the paper and inspecting the code alongside it :)

I tested it on some toy datasets and tried creating flat clusters from the cluster tree using a distance threshold instead of specifying a value for K, and started to wonder if there was a way to compute the highest possible F1 score of pairwise precision/recall for a given tree. I came up with the code below which seems to work when the dataset satisfies separability and I get an F1 score of 1.0, but i'm not sure if it holds for more practical cases.
```
def compute_highest_possible_f1(root, collapsibles):
    ''' Collapses the tree at each collapsible node ordered by min distance,
    and computes the F1 score each time until no collapsible nodes are left

    Returns the max of all computed f1 scores - this is the highest possible f1
    score achievable for this tree
    '''
    f1_scores = []
    while len(collapsibles) > 0:    
        # get the next collapsible node
        min_d, node = heappop(collapsibles)
        while not node.valid_collapse():
            try:
                min_d, node = heappop(collapsibles)
            except:
                # fresh out of collapsibles
                return root, f1_scores
            
        node.collapse()
        # compute f1 scores at each possible collapse point
        cluster_labels, gt_labels = get_clusters(root)
        f1_score = compute_f1_score(gt_labels, cluster_labels)
        f1_scores.append(f1_score)
       
        if node.parent and node.parent.valid_collapse():
            heappush(collapsibles, (node.parent.children_min_d, node.parent))
    return np.max(f1_scores)

def get_clusters(root):
    cluster_labels = []
    gt_labels = []
    idx = 0
    c_idx = 0
    flat_clusters = root.clusters()
    for c in flat_clusters:
        pts = [pt for l in c.leaves() for pt in l.pts]
        for pt in pts:
            cluster_labels.append(c_idx)
            gt_labels.append(pt[1])
        c_idx += 1
    return cluster_labels, gt_labels
```
If there's a part of your code somewhere that effectively does this and I've missed it, apologies!

**Comment by akobre01** [comment]
Glad you like the algorithm!

WRT your code: I think your code will work (provably) for separated data. But, for non-separated data I don't think it will provably work since you're collapsing nodes by order of min_d, and by doing so, you're not actually computing the F-score of _every_ tree consistent partition, but rather a subset. As written, it's not clear to me that the partitions you're skipping don't include the highest scoring partition (by F-score).

Also, I don't think we have something like this in the current repo.

**Comment by meteora9479** [comment]
Hi @avasbr, 

I am also working on creating flat clusters from the cluster tree using a distance threshold. Could you provide your code for my reference (e.g.  toy dataset(separated multi gaussian clusters) testing)?  

Thanks

**Comment by akobre01** [comment]
Hi @meteora9479,
In case it's helpful, the xcluster code is also able to pick a flat clustering from a Perch tree. Here is the beginning of the relevant python code (similar code exists in the Scala implementation): https://github.com/iesl/xcluster/blob/b08da9389523c745b0d223e3a2aed2c8bb422ec0/src/python/xcluster/eval/eval_dataset.py#L128

**Comment by avasbr** [comment]
@meteora9479 sorry for the late response, i didn't do anything special - just used sklearn blobs, pretty much exactly like this: https://scikit-learn.org/stable/auto_examples/datasets/plot_random_dataset.html#sphx-glr-auto-examples-datasets-plot-random-dataset-py 

@akobre01 that makes sense, thanks for taking a look at that. do you have any general recommendations for what strategy would work well with distance thresholds? in most real world applications, the distance threshold is more often the heuristic that is well known, e.g. distances between embeddings extracted from deep nets are well understood, whereas K is something that just comes out when a proper dist threshold is chosen.

**Comment by akobre01** [comment]
@avasbr I'm not sure I understand your question. Are you asking for an approach to pick a distance threshold by which you will create a flat clustering? If so, will you threshold the diagonal of the bounding box stored at each PNode?  In any case, we've had some luck with tuning such a threshold on a development set, i.e., choose a small subset of data, build a tree, find the best threshold wrt flat clusterings in the small tree, then use the same threshold on the full dataset.

**Comment by avasbr** [comment]
@akobre01 sorry for my lack of clarity; i meant more along the lines of if I had a distance threshold in hand (computed from a dev set, let's say), what strategy I would actually use in collapsing the tree down to flat clusters. For example, would it be reasonable to start at the root and traverse in a BFS style and collapse nodes if the stored min dist at that node exceeds the distance threshold? 

The reason I ask is that I have several datasets of varying sizes (5 - 150K points of dimensionality 128 - it's a comp. vision application where i compute embeddings on images) that I have ground truth labels for, and they're quite separable; 0.99+ on F1 score even when using the simple strategy of adding points one by one and merging or creating clusters based on distance, but when I try PERCH on these sets with exact search and the min dist strategy i mentioned above, i'm not able to replicate these numbers so I'm fairly sure I'm not using the algorithm correctly.

**Comment by akobre01** [comment]
@avasbr if you're working with the xcluster code, then you probably want to collapse the nodes whose parent's max distance is less than the threshold. The remaining leaves, after collapsing, are the roots of your clusters.

**Comment by avasbr** [comment]
@akobre01 thanks, i'll give that a shot :)
